### PR TITLE
virt-api: introduce a more relaxed admission policy for a VM that is under rename process

### LINKED ISSUE #3974: VirtualMachine VM rename VM update should succeed
/kind bug
/triage build-officer

**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind enhancement

**What happened**:
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/3883/pull-kubevirt-e2e-k8s-1.18/1292970397448278016

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by enp0s3**
Allow updating the metadata and status sections of a VM object that is under rename process.
The same admit policy approach is also seen in VM snapshot admission control.

Signed-off-by: Igor Bezukh <ibezukh@redhat.com>

**What this PR does / why we need it**:
There are some cases when VM object metadata should be updated by the controller.
For example, create the API resource version labels on a new VM object.

Fixes #3974

```release-note
None
```

### CODE REVIEW

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 1]
```diff
@@ -284,10 +284,13 @@ func validateStateChangeRequests(ar *v1beta1.AdmissionRequest, vm *v1.VirtualMac
 		}
 
 		if getRenameRequest(existingVM) != nil {
-			return []metav1.StatusCause{{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: "Modifying a VM during a rename process is not allowed",
-			}}
+			if !reflect.DeepEqual(existingVM.Spec, vm.Spec) {
```
Comment (ashleyschuett) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:287]: I think the following might read better?
```
if getRenameRequest(existingVM) != nil && !reflect.DeepEqual(existingVM.Spec, vm.Spec) {
...
}
```

[the remark this answers is not in the document]
Comment (enp0s3) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: @xpivarc You are correct. We can do so, but do you have some scenario that will justify the restriction of the update only for service accounts such as virt-controller? if so, I will have to update other admission as well i.e. VM snapshot

[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: What happens if the user adds labels while VM is being renamed? I don't think its something major but there could be undesired behavior.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 2]
```diff
@@ -284,10 +284,22 @@ func validateStateChangeRequests(ar *v1beta1.AdmissionRequest, vm *v1.VirtualMac
 		}
 
 		if getRenameRequest(existingVM) != nil {
-			return []metav1.StatusCause{{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: "Modifying a VM during a rename process is not allowed",
-			}}
+			allowed := webhooks.GetAllowedServiceAccounts()
+			if _, ok := allowed[ar.UserInfo.Username]; ok {
```
Comment (xpivarc) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:288]: shouldn't `ok` -> `!ok`

Comment (enp0s3) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:288]: @xpivarc If it's ok then the user is a service account. For a service account we allow to change the metadata and status areas. Even a service account cannot change the spec during rename process.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 3]
```diff
@@ -284,10 +284,22 @@ func validateStateChangeRequests(ar *v1beta1.AdmissionRequest, vm *v1.VirtualMac
...
+						Type:    metav1.CauseTypeFieldValueNotSupported,
+						Message: fmt.Sprint("Cannot update VM spec until rename process completes"),
+						Field:   k8sfield.NewPath("spec").String(),
+					}}
+
+				}
+			} else {
+				return []metav1.StatusCause{{
+					Type:    metav1.CauseTypeFieldValueNotSupported,
+					Message: fmt.Sprint("Modifying a VM during a rename process is not allowed"),
```
Comment (danielBelenky) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:300]: The error message here doesn't seem right to me. This error is returned if user which is not allowed to modify kubevirt components is trying to do so.

**Review by danielBelenky — CHANGES_REQUESTED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go (hunk 3 above)
Comment (enp0s3) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: @danielBelenky WDYT about the new message?

**Pull request merged**